Sapling support for persisting wallet to disk

mdr0id opened the pull request:
Closes #3388. Closes #3061.

str4d commented on src/wallet/wallet.cpp:
```diff
@@ -329,7 +346,21 @@ bool CWallet::LoadCryptedZKey(const libzcash::SproutPaymentAddress &addr, const
     return CCryptoKeyStore::AddCryptedSproutSpendingKey(addr, rk, vchCryptedSecret);
 }
-
+bool LoadCryptedSaplingZKey(const libzcash::SaplingFullViewingKey &fvk, const std::vector<unsigned char> &vchCryptedSecret)//, const boost::optional<libzcash::SaplingPaymentAddress> &defaultAddr = boost::none)
```
Missing `CWallet::`

str4d commented on src/wallet/wallet.cpp:
```diff
@@ -1976,26 +1975,24 @@ bool CWallet::SetCryptedHDSeed(const uint256& seedFp, const std::vector<unsigned
 void CWallet::SetHDChain(const CHDChain& chain, bool memonly)
 {
     LOCK(cs_wallet);
-    /* TODO: Uncomment during PR for #3388
```
Sorry, I should have been more explicit in my comment. Please use `git revert b7f9a7ae02ffcd4bce3df9e7c9e231b0d64ed0d4` to undo this change, so that all concurrent changes are also reverted.

str4d commented on src/wallet/walletdb.cpp:
```diff
@@ -124,7 +124,26 @@ bool CWalletDB::WriteCryptedZKey(const libzcash::SproutPaymentAddress & addr,
     }
     return true;
 }
+bool CWalletDB::WriteCryptedSaplingZKey(const boost::optional<libzcash::SaplingPaymentAddress> &defaultAddr,
```
`defaultAddr` is not used in this function. Remove the parameter.

str4d commented on src/wallet/walletdb.cpp:
```diff
@@ -124,7 +124,26 @@ bool CWalletDB::WriteCryptedZKey(const libzcash::SproutPaymentAddress & addr,
+    nWalletDBUpdated++;
 
+    std:cout<<"\n\nWrite Crypted Sap Z Key"<<std::endl;
+    if (!Write(std::make_pair(std::string("sapzkeymeta"), fvk.in_viewing_key()), keyMeta))
```
Missing corresponding reader in `ReadKeyValue()`.

str4d commented on src/wallet/walletdb.cpp:
```diff
@@ -124,7 +124,26 @@ bool CWalletDB::WriteCryptedZKey(const libzcash::SproutPaymentAddress & addr,
+    if (!Write(std::make_pair(std::string("sapzkeymeta"), fvk.in_viewing_key()), keyMeta))
+        return false;
+
+    if (!Write(std::make_pair(std::string("csapzkey"), fvk.in_viewing_key()), std::make_pair(fvk, vchCryptedSecret), false))
```
Missing corresponding reader in `ReadKeyValue()`.

str4d commented on src/wallet/wallet.cpp:
```diff
@@ -161,8 +161,12 @@ bool CWallet::AddSaplingZKey(
         return true;
     }
-
+    
```
Whitespace

str4d commented:
Also we need to serialize every entry in `mapSaplingIncomingViewingKeys`; easiest to do this when we add to that map, which is when we add a new spending key. This ties into #3511 which @Eirik0 is working on, where IIRC he is adding functions to the wallet / keystore stack to enable adding to this map outside of when we add a new spending key.

bitcartel commented on src/wallet/wallet.h:
```diff
@@ -509,8 +519,7 @@ class CWalletTx : public CMerkleTx
         READWRITE(fSpent);
-        // TODO:
-        //READWRITE(mapSaplingNoteData);
+        READWRITE(mapSaplingNoteData);
```
git grep vUnused
wallet/wallet.h:503:        std::vector<CMerkleTx> vUnused; //! Used to be vtxPrev
wallet/wallet.h:504:        READWRITE(vUnused);

mdr0id commented:
@str4d Addressed comments and awaiting another pass to ensure I'm heading in the right direction. I will take a peek at #3511, to see if it fills in the gap of where I'm stuck.

mdr0id commented:
@zkbot try

mdr0id commented:
@zkbot try

mdr0id commented:
@zkbot try

daira commented on src/wallet/wallet.cpp:
```diff
@@ -167,6 +167,10 @@ bool CWallet::AddSaplingZKey(
     }
 
     // TODO: Persist to disk
```
Is this comment stale?

mdr0id replied:
Yes, it was used as a place holder for code implementation. I will clean this up for final pass.

daira commented on src/wallet/wallet.cpp:
```diff
@@ -303,6 +307,16 @@ bool CWallet::AddCryptedSaplingSpendingKey(const libzcash::SaplingFullViewingKey
         return true;
     {
         // TODO: Sapling - Write to disk
```
Is this comment stale?

mdr0id replied:
Yes, it was used a place holder for code implementation. I will clean this up for final pass.

daira commented on src/wallet/wallet.h:
```diff
@@ -509,8 +523,11 @@ class CWalletTx : public CMerkleTx
+        if (fOverwintered && nVersion >= SAPLING_TX_VERSION) 
+        {
+            READWRITE(mapSaplingNoteData);                
+        }
```
Did we decide what to do about records written by 2.0.0 after Sapling activation, and then read by 2.0.1? Has that case been smoke-tested?

mdr0id replied:
There was some discussion around this, but I don't recall the final outcome. This has not been smoke tested as of 10/2/2018

str4d replied:
Based on looking at the `LoadWallet` logic in `src/wallet/walletdb.cpp`, what @mdr0id and I concluded was:

- The Sapling v4 transactions written by 2.0.0 would be read by 2.0.1, and the `READWRITE` above would throw an error due to insufficient data.
- This gets caught by the loading logic, where there is a special case for transaction read fails that schedules a from-genesis rescan.
- The rescan will load the transactions directly into memory from the block data, and then immediately write them to disk.
- As the transaction ID is used as the database key, the 2.0.1 write overwrites the 2.0.0 record, fixing it for the next time the node starts.

We have not tested this AFAIK.

bitcartel replied:
Need to sync 2.0.0 with testnet, tap the testnet faucet to get some funds (it will be v4 transaction), shut down, and launch node with this PR

bitcartel replied:
I can confirm the rescan logic is activated as we initially discussed and later concluded.

I synced up a wallet with testnet using v2.0.0 which had a v4 Sapling transaction (transparent value transfer) in the wallet.

Switching to 2.0.1 (this PR) triggered a rescan.

Switching back to 2.0.0 did not trigger a rescan (as expected, since the mapSaplingNoteData is appended to the end of the serialized data expected by 2.0.0)

Switching back to 2.0.1 (this PR) triggers a rescan.

When triggering the rescan, `debug.log` output looks like this:

```
2018-10-06 02:05:50 init message: Loading wallet...
2018-10-06 02:05:50 Warning: Warning: error reading wallet.dat! All keys read correctly, but transaction data or ad
dress book entries might be missing or incorrect.
2018-10-06 02:05:50  wallet                   31ms
2018-10-06 02:05:50 init message: Rescanning...
2018-10-06 02:05:50 Rescanning last 299187 blocks (from block 0)...
2018-10-06 02:05:50 AddToWallet a2f121c7219ff17b68de66963b4bb5220e372687076067d90717e4d670c2e2ee  new
...
...
2018-10-06 02:05:55 AddToWallet dda3941da3ab77f3bdc61728354e6e9285c31e456f38b4cc39278ff5c3e7d0fd  new
2018-10-06 02:06:00  rescan                 9629ms
```

daira commented on src/wallet/walletdb.cpp:
```diff
@@ -141,6 +158,18 @@ bool CWalletDB::WriteZKey(const libzcash::SproutPaymentAddress& addr, const libz
+{
+    nWalletDBUpdated++;
+    
+    if (!Write(std::make_pair(std::string("sapzkmeta"), ivk), keyMeta))
```
Should this be "sapzkeymeta"? https://github.com/zcash/zcash/pull/3517/files#diff-e506682270036e4c8e0b8159a34b652dR715

If it should then this may indicate lack of test coverage.

mdr0id replied:
Yes, some of the naming conventions weren't very clear, so I missed this one. Will update for final pass.

mdr0id commented:
@zkbot try

bitcartel commented:
Used z_importkey but could not send, got error: "Missing witness for Sapling note"
```
./zcash-cli z_getbalance ztestsapling1jesyn5p3vnn9wy3pcp7gvtudd6g3q646n0rgvctgmjmvvatm7fkejsvpjldugypw6rrzwf0mcfe
4.18312000

./zcash-cli z_sendmany ztestsapling1jesyn5p3vnn9wy3pcp7gvtudd6g3q646n0rgvctgmjmvvatm7fkejsvpjldugypw6rrzwf0mcfe '[{"address":"ztestsapling1jesyn5p3vnn9wy3pcp7gvtudd6g3q646n0rgvctgmjmvvatm7fkejsvpjldugypw6rrzwf0mcfe", "amount":4.1}]'
opid-7346f483-9ed1-45ad-919a-2c93062bc59a

./zcash-cli z_getoperationresult
[
  {
    "id": "opid-7346f483-9ed1-45ad-919a-2c93062bc59a",
    "status": "failed",
    "creation_time": 1538596812,
    "error": {
      "code": -4,
      "message": "Missing witness for Sapling note"
    },

```

mdr0id commented:
Will sanity check that #3560 addresses above mentioned issue, once it is merged into master.

mdr0id commented:
@zkbot try

Eirik0 commented on src/keystore.cpp:
```diff
@@ -171,7 +171,7 @@ bool CBasicKeyStore::AddSaplingIncomingViewingKey(
 
     // Add addr -> SaplingIncomingViewing to SaplingIncomingViewingKeyMap
     mapSaplingIncomingViewingKeys[addr] = ivk;
```
I think this still needs to be persisted in some way.

mdr0id replied:
I can't recall the history/implementation flow here. @str4d might be able to shed some light on this

str4d replied:
See my review comment.

Eirik0 commented on src/wallet/wallet.cpp:
```diff
@@ -168,7 +168,7 @@ bool CWallet::AddSaplingZKey(
     if (!IsCrypted()) {
         auto ivk = sk.expsk.full_viewing_key().in_viewing_key();
-        return CWalletDB(strWalletFile).WriteSaplingZKey(ivk,sk, mapSaplingZKeyMetadata[ivk],defaultAddr); 
+        return CWalletDB(strWalletFile).WriteSaplingZKey(defaultAddr.get(),ivk,sk, mapSaplingZKeyMetadata[ivk]); 
```
defaultAddr is no longer optional and the .get() needs to be removed from this method.
Edit: and similar methods

Eirik0 commented on qa/rpc-tests/wallet_persistence.py:
```diff
@@ -0,0 +1,32 @@
+from test_framework.util import assert_true, start_nodes, stop_nodes, \
+    wait_bitcoinds
+
+class WalletPersistenceTest (BitcoinTestFramework):
```
It would be good to update this test to make sure that we can still spend a sapling note after having restarted.

mdr0id replied:
I thought there was a test that does this, but I need to confirm. We should at least put a note in one of the test that assumes this test ensures a disk persists is valid.

bitcartel replied:
Pushed a commit to update the test.

mdr0id commented:
👎 I'm not convinced Homu tells the truth. Per suggestion, going to re PR this branch and toss it at Homu

mdr0id commented:
@zkbot try

mdr0id commented on src/wallet/walletdb.cpp:
```diff
@@ -809,7 +885,10 @@ DBErrors CWalletDB::LoadWallet(CWallet* pwallet)
                 // we assume the user can live with:
                 if (IsKeyType(strType))
+                {
+                    std::cout<<"IN IS KEYTYPE "<<strType<<"   "<<strErr<<std::endl;
```
This was from me forgetting to revert it late at night. Will remove

str4d replied:
Yes, remove this line or turn it into an intentional error (though we do print an error in `init.cpp` where the `DB_CORRUPT` eventually gets raised to).

mdr0id commented:
@zkbot try

str4d commented on src/wallet/walletdb.cpp:
```diff
@@ -124,7 +124,25 @@ bool CWalletDB::WriteCryptedZKey(const libzcash::SproutPaymentAddress & addr,
+    const bool fEraseUnencryptedKey = true;
+    nWalletDBUpdated++;
 
+    if (!Write(std::make_pair(std::string("sapzkeymeta"), addr), keyMeta))
```
This is incorrectly using `addr` in the key, while the read side correctly expects `ivk`.

str4d commented on src/wallet/walletdb.cpp:
```diff
@@ -141,6 +159,18 @@ bool CWalletDB::WriteZKey(const libzcash::SproutPaymentAddress& addr, const libz
+{
+    nWalletDBUpdated++;
+    
+    if (!Write(std::make_pair(std::string("sapzkeymeta"), defaultAddr ), keyMeta))
```
This is incorrectly using `defaultAddr` in the key, while the read side correctly expects `ivk`.

str4d commented on src/wallet/walletdb.cpp:
```diff
@@ -511,6 +541,23 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
+        else if (strType == "sapzkey")
+        {
+            libzcash::SaplingPaymentAddress addr;
+            ssKey >> addr;
```
```
libzcash::SaplingIncomingViewingKey ivk;
ssKey >> ivk;
```

str4d commented on src/wallet/walletdb.cpp:
```diff
@@ -624,6 +671,23 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
+        else if (strType == "csapzkey")
+        {
+            libzcash::SaplingPaymentAddress addr;
+            ssKey >> addr;
```
```
libzcash::SaplingIncomingViewingKey ivk;
ssKey >> ivk;
```

str4d commented on src/wallet/walletdb.cpp:
```diff
@@ -141,6 +159,18 @@ bool CWalletDB::WriteZKey(const libzcash::SproutPaymentAddress& addr, const libz
+    if (!Write(std::make_pair(std::string("sapzkeymeta"), defaultAddr ), keyMeta))
+        return false;
+
+    return Write(std::make_pair(std::string("sapzkey"), defaultAddr), key, false);
```
`return Write(std::make_pair(std::string("sapzkey"), ivk), key, false);`

str4d commented on src/wallet/walletdb.cpp:
```diff
@@ -124,7 +124,25 @@ bool CWalletDB::WriteCryptedZKey(const libzcash::SproutPaymentAddress & addr,
+    if (!Write(std::make_pair(std::string("sapzkeymeta"), addr), keyMeta))
+        return false;
+
+    if (!Write(std::make_pair(std::string("csapzkey"), addr), std::make_pair(fvk, vchCryptedSecret), false))
```
`if (!Write(std::make_pair(std::string("csapzkey"), ivk), std::make_pair(fvk, vchCryptedSecret), false))` (and get `ivk` from `fvk`).

str4d commented on src/wallet/walletdb.cpp:
```diff
@@ -124,7 +124,25 @@ bool CWalletDB::WriteCryptedZKey(const libzcash::SproutPaymentAddress & addr,
+        return false;
+    if (fEraseUnencryptedKey)
+    {
+        Erase(std::make_pair(std::string("sapzkey"), addr));
```
`Erase(std::make_pair(std::string("sapzkey"), ivk));`

str4d commented on src/wallet/walletdb.cpp:
```diff
@@ -511,6 +541,23 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
+            libzcash::SaplingExtendedSpendingKey key;
+            ssValue >> key;
+
+            if (!pwallet->LoadSaplingZKey(key, addr))
```
Argh, this is an example where we won't know an address to insert here (because that data will be stored in a different record), so we *do* need the second parameter to be optional. Sorry @Eirik0 :( Hopefully we can just revert that commit without too much trouble.

str4d commented on src/wallet/walletdb.cpp:
```diff
@@ -624,6 +671,23 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
+            ssValue >> vchCryptedSecret;
+            wss.nCKeys++;
+
+            if (!pwallet->LoadCryptedSaplingZKey(fvk ,vchCryptedSecret, addr ))
```
Likewise, we won't know `addr` here, so the last parameter needs to be optional. Also whitespace.

str4d commented on src/keystore.cpp:
```diff
@@ -157,7 +157,7 @@ bool CBasicKeyStore::AddSaplingFullViewingKey(
     mapSaplingFullViewingKeys[ivk] = fvk;
 
-    return AddSaplingIncomingViewingKey(ivk, defaultAddr);
+    return CBasicKeyStore::AddSaplingIncomingViewingKey(ivk, defaultAddr);
```
I believe the deadlock was caused by the fact that `CWalletDB::LoadWallet` when loading a Sapling key calls `LoadSaplingZKey`, which calls `CCryptoKeyStore::AddSaplingSpendingKey`, which ends up calling here. If we call `AddSaplingIncomingViewingKey` here then the wallet ends up trying to write the result out, which likely causes a deadlock in the underlying database.

The effect of this change is that when creating a new Sapling key, the `defaultAddr -> ivk` map entry is not persisted to disk (while other `addr -> ivk` maps are, because they call `AddSaplingIncomingViewingKey` directly). However, this is fine, because we can always recover the default address from the main key entry (either unencrypted from the extended spending key, or encrypted via the extended full viewing key).

Eirik0 commented:
ACK

daira commented on src/wallet/walletdb.cpp:
```diff
@@ -911,11 +1015,20 @@ DBErrors CWalletDB::FindWalletTx(CWallet* pwallet, vector<uint256>& vTxHash, vec
+                } catch (...) {
+                    // Decode failure likely due to Sapling v4 transaction format change
+                    // between 2.0.0 and 2.0.1. As user is requesting deletion, log the
+                    // transaction entry and then mark it for deletion anyway.
```
This is just `FindWalletTx`, so why did we conclude that the user is requesting deletion?

str4d replied:
This code was added by upstream when adding `-zapwallettxes`, and is only called from `CWalletDB::ZapWalletTx`.

daira replied:
Add a comment at the top of the function to that effect. Or rename to `FindWalletTxToZap`.

str4d replied:
Opened #3572.

Eirik0 commented:
reACK

daira commented on src/wallet/gtest/test_wallet_zkeys.cpp:
```diff
@@ -288,7 +331,6 @@ TEST(wallet_zkeys_tests, WriteViewingKeyDirectToDB) {
  * This test covers methods on CWalletDB to load/save crypted z keys.
  */
-/* TODO: Uncomment during PR for #3388
 TEST(wallet_zkeys_tests, write_cryptedzkey_direct_to_db) {
```
Naming should be CamelCase.

str4d replied:
Opened #3575.

daira commented on src/wallet/wallet.cpp:
```diff
@@ -166,11 +166,36 @@ bool CWallet::AddSaplingZKey(
-    // TODO: Persist to disk
+    if (!IsCrypted()) {
+        auto ivk = sk.expsk.full_viewing_key().in_viewing_key();
+        return CWalletDB(strWalletFile).WriteSaplingZKey(ivk, sk, mapSaplingZKeyMetadata[ivk]);
```
`mapSaplingZKeyMetadata` has type `std::map<libzcash::SaplingPaymentAddress, CKeyMetadata>`, and if the key doesn't exist in the map, `std::map`'s `operator[]` uses the default constructor of the target type to insert a new instance. In this case it will be a `CKeyMetadata` instance with the current version and an "unknown" time.

I'm unsure whether that was intended or whether this is a case that shouldn't ever happen.

str4d replied:
This is, I believe, expected behaviour:

- When calling `AddSaplingZKey` from `GenerateNewSaplingZKey`, we set `mapSaplingZKeyMetadata` there first.
- When calling `AddSaplingZKey` from `AddSpendingKeyToWallet`, we set `mapSaplingZKeyMetadata` appropriately afterwards.

daira commented on src/wallet/walletdb.cpp:
```diff
@@ -511,6 +554,23 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
+                return false;
+            }
+
+            //add checks for integrity
```
Should be a TODO to make the comment easier to find, and/or file a ticket. Or remove if stale.

str4d replied:
Opened #3576.

daira commented on src/wallet/walletdb.cpp:
```diff
@@ -651,6 +728,32 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
+
+            wss.nZKeyMeta++;
+
+            pwallet->LoadSaplingZKeyMetadata(ivk, keyMeta);
```
Currently `LoadSaplingZKeyMetadata` always returns `true` as noted by @bitcartel, but it does return a boolean, so this is a maintenance hazard. Please file a ticket.

str4d replied:
Opened #3571.

str4d commented on src/wallet/crypter.cpp:
```diff
@@ -465,12 +465,12 @@ bool CCryptoKeyStore::AddSaplingSpendingKey(
-        auto fvk = sk.expsk.full_viewing_key();
-        if (!EncryptSecret(vMasterKey, vchSecret, fvk.GetFingerprint(), vchCryptedSecret)) {
+        auto extfvk = sk.ToXFVK();
+        if (!EncryptSecret(vMasterKey, vchSecret, extfvk.fvk.GetFingerprint(), vchCryptedSecret)) {
```
As wallet encryption is still an experimental feature, we have latitude to change how we do this later if we need to. We just need something self-consistent for now.

str4d replied:
Opened #3573.

str4d commented:
Opened new issues for the remaining non-blocking comments.

@zkbot r+
